### PR TITLE
Delete the SolutionExplorerShim assembly and merge it into our VisualStudio.Implementation project

### PR DESCRIPTION
**Pull request by jasonmalinowski**
We add the Analyzers node and related nodes to the Solution Explorer via an API that was exposed via MEF. Because there was
no way to control which projects loaded us, we loaded _always_, even in solutions that didn't contain C# or VB. To limit the damage, we had a separate assembly that carefully avoided ever loading C# or VB unless we actually had a project around.

The shell has updated the Solution Explorer API to allow us to specify as MEF metadata which projects we should apply to. This allows us to only be created when necessary, which means we can delete the entire assembly and merge it back in. We can also delete a bunch of lazy loading hacks and generally clean all sorts of things up.

Commit at a time is _highly_ recommended here.

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
I suspect @mavasani and @tmeschter will be happy to see this project and the lazy loading hacks going away.

File: src/VisualStudio/Core/Impl/SolutionExplorer/AnalyzersFolderItem/AnalyzersFolderItemProvider.cs [hunk 1]
```diff
@@ -24,37 +19,24 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplore
     [Export(typeof(IAttachedCollectionSourceProvider))]
     [Name("AnalyzersFolderProvider")]
     [Order(Before = HierarchyItemsProviderNames.Contains)]
+    [AppliesToProject("(CSharp | VisualBasic) & !CPS")] // in the CPS case, the Analyzers folder is created by the project system
```
Comment (mavasani) [inline src/VisualStudio/Core/Impl/SolutionExplorer/AnalyzersFolderItem/AnalyzersFolderItemProvider.cs:22]: Nit: Define a string constant given it is used more than once?

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/BaseDiagnosticItem.BrowseObject.cs [hunk 2]
```diff
@@ -56,7 +53,7 @@ public string HelpLink
             {
                 get
                 {
-                    return _diagnosticItem.GetHelpLink()?.AbsoluteUri;
+                    return _diagnosticItem.GetHelpLink()?.AbsoluteUri!;
```
Comment (mavasani) [inline src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/BaseDiagnosticItem.BrowseObject.cs:56]: Interesting syntax. Should this be `_diagnosticItem.GetHelpLink().AbsoluteUri!` if we know the result of the expression cannot be null?

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItem.cs [hunk 3]
```diff
@@ -21,7 +19,17 @@ public CpsDiagnosticItem(CpsDiagnosticItemSource source, DiagnosticDescriptor de
         }
 
         public override ProjectId ProjectId => _source.ProjectId;
-        protected override AnalyzerReference AnalyzerReference => _source.AnalyzerReference;
+        protected override AnalyzerReference AnalyzerReference
+        {
+            get
+            {
+                // _source.AnalyzerReference can be null if the source hasn't found it's reference yet;
```
Comment (mavasani) [inline src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItem.cs:26]: Should we have a `Debug.Assert(_source.AnalyzerReference != null);` to ensure the contract is not violated in future?

**Review by mavasani — APPROVED** [review]
Looks great! Glad to see this project going away...

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/BaseDiagnosticItem.BrowseObject.cs (hunk 2 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/BaseDiagnosticItem.BrowseObject.cs:56]: Alright I must have misread something here -- I had thought I was suppressing AbsoluteUri being null but clearly the ? doesn't make sense either. I'll...inspect further. :smile:

File: src/VisualStudio/Core/Impl/SolutionExplorer/AnalyzersFolderItem/AnalyzersFolderItemProvider.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Impl/SolutionExplorer/AnalyzersFolderItem/AnalyzersFolderItemProvider.cs:22]: I thought about it, but I wasn't sure anything would be really clearer than just seeing the literal expression. And even if somebody _did_ have to change this, I imagine they'd have to re audit all of the other ones to see if they still made sense, so this isn't really a good case where a constant means you can change it in one place and not have to worry about bugs you introduced.

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/BaseDiagnosticItem.BrowseObject.cs (hunk 2 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/BaseDiagnosticItem.BrowseObject.cs]: Alright I couldn't read a tooltip. Fixed.

File: src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItem.cs (hunk 3 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItem.cs:27]: Added a Contract.ThrowIfNull().

**Pull request merged**